pkg/kapp/app/recorded_app.go

```diff
@@ -196,27 +191,88 @@ func (a *RecordedApp) createOrUpdate(c *corev1.ConfigMap, labels map[string]stri
 				return fmt.Errorf("Getting app: %s", err)
 			}
 
-			err = a.mergeAppUpdates(existingConfigMap, labels)
-			if err != nil {
-				return err
+			return a.updateApp(existingConfigMap, labels)
+		}
+
+		return fmt.Errorf("Creating app: %s", err)
+	}
+
+	return nil
+}
+
+func (a *RecordedApp) updateApp(existingConfigMap *corev1.ConfigMap, labels map[string]string) error {
+	err := a.mergeAppUpdates(existingConfigMap, labels)
+	if err != nil {
+		return err
+	}
+
+	_, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Update(context.TODO(), existingConfigMap, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("Updating app: %s", err)
+	}
+
+	return nil
+}
+
+func (a *RecordedApp) RenamePrevApp(prevAppName string, labels map[string]string) error {
+	defer a.logger.DebugFunc("RenamePrevApp").Finish()
+	var c *corev1.ConfigMap
+	var err error
+
+	if a.isMigrationEnabled() {
+		c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.fqName, metav1.GetOptions{})
+		if err == nil {
+			return a.updateApp(c, labels)
+		} else if err != nil {
+			if errors.IsNotFound(err) {
+				c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.name, metav1.GetOptions{})
+				if err == nil {
+					return a.migrate(c, labels, a.fqName)
+				}
 			}
+		}
+	} else {
+		c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), a.name, metav1.GetOptions{})
+		if err == nil {
+			return a.updateApp(c, labels)
+		}
+	}
 
-			_, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Update(context.TODO(), existingConfigMap, metav1.UpdateOptions{})
-			if err != nil {
-				return fmt.Errorf("Updating app: %s", err)
+	if errors.IsNotFound(err) {
+		if a.isMigrationEnabled() {
+			c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName+AppSuffix, metav1.GetOptions{})
+			if err == nil {
+				return a.migrate(c, labels, a.fqName)
+			} else if err != nil {
+				if errors.IsNotFound(err) {
+					c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName, metav1.GetOptions{})
+					if err == nil {
+						return a.migrate(c, labels, a.fqName)
+					}
+				}
 			}
+		} else {
+			c, err = a.coreClient.CoreV1().ConfigMaps(a.nsName).Get(context.TODO(), prevAppName, metav1.GetOptions{})
+			if err == nil {
+				err := a.renameConfigMap(c, a.name, a.nsName)
+				if err != nil {
+					return err
+				}
 
-			return nil
+				return a.updateApp(c, labels)
+			}
 		}
 
-		return fmt.Errorf("Creating app: %s", err)
+		if errors.IsNotFound(err) {
+			return a.CreateOrUpdate(labels)
```

Comment:
I have been wondering about this a bit. This will result in a lot of duplicate calls when both the app doesn't exist. But I guess that's okay to have as it will only happen once when the app is deployed for the first time (vs we will have to create the configmap here itself based on the value of isMigrationEnabled which will result in code duplication).

Reply:
Right, I think most of this code introduces a potential for lots of calls in the worst case. But it's optimized for the case of I have an app already running in the cluster. I think the hit is ok also because it's only the first time. And tbh I am not really sure how we can avoid these duplicate calls given the stateless nature of kapp.

Reply:
I just remembered something. We should ideally *not* use a public method inside another public method. I don't think it's a rule but rather a good to have thing. We are already calling `CreateOrUpdate` in deploy after `RenamePrevApp` (although in an else), maybe we could somehow use that to do our job?

(I know I have already approved the PR, but this is something that struck my mind while working on something else and thought of sharing 😅 )

Reply:
Cool, I updated the code. It makes things a little longer in that method, but I think it's a little more readable also.